worker/src/core/loaders/message-worker/index.ts

```diff
@@ -248,6 +256,111 @@ const enqueueAndSend = async (): Promise<void> => {
+  void connection
+    .query(
+      "SELECT json_build_object('id', c.id," +
+        "'campaign_name', c.name," +
+        "'created_at', c.created_at," +
+        "'unsent_count', s.unsent," +
+        "'error_count', s.errored," +
+        "'sent_count', s.sent," +
+        "'invalid_count', s.invalid," +
+        "'notification_email', u.email," +
+        "'halted', c.halted) as result FROM campaigns c INNER JOIN statistics s ON c.id=s.campaign_id INNER JOIN users u ON c.user_id=u.id WHERE c.id=:campaign_id_input;",
+      {
+        replacements: { campaign_id_input: campaignId },
+        type: QueryTypes.SELECT,
+      }
+    )
+    .then(async (result) => {
+      const jsonRes = get(result, '[0].result') || {}
+      const {
+        id: campaignId,
+        campaign_name: campaignName,
+        created_at: createdAt,
+        unsent_count: unsentCount,
+        error_count: errorCount,
+        sent_count: sentCount,
+        invalid_count: invalidCount,
+        notification_email: notificationEmail,
+        halted: halted,
+      } = jsonRes
+
+      let mail: MailToSend | void
+      if (campaignId) {
+        // craft and send mail here
+        // if halted, send halted notif
+        if (halted) {
+          mail =
+            await NotificationService.generateHaltedCampaignNotificationEmail(
+              client,
+              notificationEmail,
+              campaignName
+            )
+        } else {
+          // for scheduled, visible_at must be after created_at
+          // normal flow, pull out visibleAt from job_queue table.
+
+          const visibleAt = await connection
+            .query(
+              "select visible_at from job_queue where campaign_id=:campaign_id_input and status = 'LOGGED' order by created_at desc limit 1;",
+              {
+                replacements: { campaign_id_input: campaignId },
+                type: QueryTypes.SELECT,
+              }
+            )
+            .then(async (result) => {
+              return get(result, '[0].visible_at')
+            })
+          if (new Date(createdAt) < visibleAt) {
+            mail =
+              await NotificationService.generateScheduledCampaignNotificationEmail(
+                client,
+                notificationEmail,
+                campaignName,
+                unsentCount,
+                errorCount,
+                sentCount,
+                invalidCount
+              )
+          }
+        }
+        if (mail) {
+          // Send email using node mailer
+          // cannot use singleton mailclient like backend cuz worker will self destruct
+          // instantiate it only when needed
+          const mailClient = new MailClient(
+            config.get('mailOptions'),
+            config.get('mailOptions.callbackHashSecret'),
+            config.get('mailFrom'),
+            config.get('mailConfigurationSet')
+          )
+          const isEmailSent = await NotificationService.sendEmail(
```

Comment:
iinw if `isEmailSent` is null, something has gone wrong and we should throw an error? (I think the underlying method would already throw an error, but we should also do it here?)

Reply:
there may not necessarily be an error even if the mail was not successfully sent. even if there was, the function sendEmail already catches the error. 

On the side once again, this archetype is done referencing the existing code that has been already been written and tested to be working.